### PR TITLE
Parse quote characters

### PR DESCRIPTION
**Pull request by Yirmandias**
- Modification of the parse function to support quote, quasiquote and unquote characters.
- Modification of enum Token:
```rust
enum Token {
    Sym { start: usize, end: usize, start_pos: Pos },
    LParen(Pos),
    RParen(Pos),
    //new fields
    Quote(Pos),
    QuasiQuote(Pos),
    Unquote(Pos),
}
```
- Modification of the function tokenize to support new kind of Token.
- Modification of read to transform new tokens into corresponding symbols and construct and SList containing the quote expression.

### CODE REVIEW

**Comment by arbimo** [comment]
Clippy errors are likely due to the release of Rust 1.54 that has new clippy rules automatically picked up by CI. We need to address them independently.

File: planning/src/parsing/sexpr.rs [hunk 1]
```diff
@@ -339,7 +351,27 @@ fn read(tokens: &mut std::iter::Peekable<core::slice::Iter<Token>>, src: &std::s
                 }
             }
         }
+
         Some(Token::RParen(_)) => bail!("Unexpected closing parenthesis"),
+        Some(quotting) => {
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:356]: typo: quoting

File: planning/src/parsing/sexpr.rs [hunk 2]
```diff
@@ -339,7 +351,27 @@ fn read(tokens: &mut std::iter::Peekable<core::slice::Iter<Token>>, src: &std::s
...
+
+            let mut es = vec![SExpr::Atom(sym_quote)];
+            let e = read(tokens, src)?;
+            //let loc = e.loc();
+            es.push(e);
+            //Compute the span
+            Ok(SExpr::List(SList {
+                list: es,
+                source: src.clone(),
+                span: Span::new(*start, *start), //TODO: find a way to declare the span
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:372]: need to fix this and the the tests for contextual display.
My intuition is that the span should finish at the end of `e`

File: planning/src/parsing/sexpr.rs [hunk 3]
```diff
@@ -339,7 +351,27 @@ fn read(tokens: &mut std::iter::Peekable<core::slice::Iter<Token>>, src: &std::s
...
             }
         }
+
         Some(Token::RParen(_)) => bail!("Unexpected closing parenthesis"),
+        Some(quotting) => {
+            let (sym_quote, start) = match quotting {
+                Token::Quote(s) => (Sym::new("quote"), s),
+                Token::QuasiQuote(s) => (Sym::new("quasiquote"), s),
+                Token::Unquote(s) => (Sym::new("unquote"), s),
+                _ => bail!("Unexpected token, should be Quote, QuasiQuote or Unquote"),
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:361]: Should be `unreachable!(...)` : if you reach this it is a bug in our code, not in the user input.

File: planning/src/parsing/sexpr.rs (hunk 2 above)
Comment (Yirmandias) [reply planning/src/parsing/sexpr.rs:372]: The function to get the end of `e` returns a Loc object.
```
e.loc().end() -> Loc.
```
From this **Loc** we cannot get the **end** Pos

Comment (arbimo) [reply planning/src/parsing/sexpr.rs:372]: Essentially, you need `e.span.end`. 
`span` is private is this case but you can add an accessor for the start and end `Pos` of a `Loc`

File: planning/src/parsing/sexpr.rs [hunk 4]
```diff
@@ -265,39 +270,62 @@ fn tokenize(source: std::sync::Arc<Input>) -> Vec<Token> {
...
+                    } else if n == ',' {
+                        tokens.push(Token::Unquote(pos));
+                    }
                 }
             }
         } else if cur_start == None {
             cur_start = Some(index);
         }
-        index += 1;
+        //index += 1;
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:325]: remove

File: planning/src/parsing/sexpr.rs [hunk 5]
```diff
@@ -354,6 +401,26 @@ mod tests {
         assert_eq!(&formatted, output);
     }
 
+    #[test]
+    fn parsing_string() {
+        formats_as("(a \"b c\" d)", "(a \"b c\" d)");
+        formats_as("(a \"(b c)\" d)", "(a \"(b c)\" d)");
+        formats_as("(a b); \"(a b)\"", "(a b)");
+    }
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:409]: Missing test cases for `;` and `\n` in a string.

File: planning/src/parsing/sexpr.rs [hunk 6]
```diff
@@ -244,7 +247,7 @@ fn tokenize(source: std::sync::Arc<Input>) -> Vec<Token> {
     let mut tokens = Vec::new();
 
     // current index into `s`
-    let mut index = 0;
+    //let mut index = 0;
```
Comment (arbimo) [inline planning/src/parsing/sexpr.rs:250]: remove

**Review by arbimo — CHANGES_REQUESTED** [review]
Looks good to me. There are just 3 minors changes I commented on.

At this point it might be good to split the Atom between a String, a symbol and a number. This could be done in another PR however.

File: planning/src/parsing/sexpr.rs (hunk 4 above)
Comment (Yirmandias) [reply planning/src/parsing/sexpr.rs]: Resolved in 5587ae9

File: planning/src/parsing/sexpr.rs (hunk 5 above)
Comment (Yirmandias) [reply planning/src/parsing/sexpr.rs:409]: Resolved in 5587ae9

File: planning/src/parsing/sexpr.rs (hunk 6 above)
Comment (Yirmandias) [reply planning/src/parsing/sexpr.rs]: Resolved in 5587ae9

**Pull request merged**